src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -220,17 +223,41 @@ class ShenandoahRegionPartitions {
 
   inline size_t available_in(ShenandoahFreeSetPartitionId which_partition) const {
     assert (which_partition < NumPartitions, "selected free set must be valid");
-    return _capacity[int(which_partition)] - _used[int(which_partition)];
+    shenandoah_assert_heaplocked();
+    assert(_available[int(which_partition)] == _capacity[int(which_partition)] - _used[int(which_partition)],
+           "Expect available (%zu) equals capacity (%zu) - used (%zu) for partition %s",
+           _available[int(which_partition)], _capacity[int(which_partition)], _used[int(which_partition)],
+           partition_membership_name(ssize_t(which_partition)));
+    return _available[int(which_partition)];
+  }
+
+  // Acquire heap lock and return available_in, assuming heap lock is not acquired by the caller.
```

Comment:
Sorry - can we change this comment too? This method does _not_ acquire the lock in release builds. Comment could mention that it acquires the lock only for the correctness of the assertion?

Reply:
Very good point.  I've made this change.